### DISCUSSION TITLE
Username and Password in the PacketInterceptor

### DISCUSSION THREAD

**Discussion by thiagonuic** [davidepianca98/KMQTT#9] [Q&A] [answered]
Hi,

Is it possible to get the authentication information in the packet interceptor? My use case would be to validate that certain user can send information for a topic or validate the payload.

Thanks!

**Comment by davidepianca98** [comment, marked as answer]
Hello, it is possible to pass to the Broker constructor an object of type Authorization, that defines the method `fun authorize(clientId: String, topicName: String, isSubscription: Boolean): Boolean`. There you can define an authorization policy, that method gets called on publish and on subscribe (isSubscription=true). There you have the clientId and the topicName.

**Reply by thiagonuic** [reply]
Actually it would be useful to have the username and password in the packet of type MQTTPublish. With that it would be possible to validate the payload as well. Do you think it is possible and does it make sense to you?

**Reply by davidepianca98** [reply]
The username and password are discarded after the initial authentication, then the clientId identifies the authenticated client, and if the client wasn't authenticated the publish wouldn't be possible. I don't know if storing the username and password for the whole duration of the session is a great idea, but passing the payload or the MQTT5 properties is definitely doable, if you would like to store data for your custom validation there. Let me know what you think, I definitely have to think more if there is a better way of doing this.

**Reply by thiagonuic** [reply]
Having the payload in the authorization would help to validate its contents. Another option I can think of is to cache clientId and username/password in a session way. The goal here is to take the payload, username and password and forward to a REST API to process the information. So forwarding the authentication information will make it easier to validate the ownership of resources in another service. Let me know your thoughts.

**Reply by thiagonuic** [reply]
On a side note, how do I relate the clientId with the username from the authentication?

**Reply by davidepianca98** [reply]
It is interesting what you are saying about the REST API, I'll have to think about that. There must be some way of authenticating to the REST API and then identify with the clientId.
As for adding the payload in the authorization callback I can do it easily.
You can set the clientId the same as the username if you want in the connect packet, then the server won't randomly generate it (as required by the MQTT specs). The requirement is that they must be unique, but I think the username as you mean it is definitely  unique.

**Reply by davidepianca98** [reply]
So this is the enhancement I'm thinking of implementing:
`fun authorize(clientId: String, username: String?, topicName: String, isSubscription: Boolean): Boolean`
`fun packetReceived(clientId: String, username: String?, packet: MQTTPacket)`
What do you think about this? I still think saving the password isn't a good idea.

**Reply by thiagonuic** [reply]
For sure this makes it easier to handle things as for the authorization on the MQTT side. In my case, perhaps too specific, was to avoid the duplicate code to validate the payload in authorize, and keep it only in the REST API when I parse and forward the MQTTPublish intercept. So I would just forward the username and password to the REST API and it would authorize further in the processing pipeline.

- One thing missing there in the proposed change in the authorize interface is the "payload".
- One suggestion is to have perhaps a flag in the constructor of the broker whether to keep or not the password. In Spring Boot by default it does not keep username and password after Spring Security has done its thing. But there is a flag which allows you to keep those things so you can grab them in your application in case you needed. Here is the docs for reference: https://docs.spring.io/spring-security/site/docs/3.1.x/reference/springsecurity-single.html#core-services-erasing-credentials

**Reply by davidepianca98** [reply]
Ok for the payload in the authorize method.
I like the idea of optionally saving the password, I'd put an additional parameter in the broker constructor that's by default set to false, so a user has to explicitly set it to true.
I will do the modifications as soon as possible (I hope during the weekend) and report here when it's done.

**Reply by thiagonuic** [reply]
Thanks for being responsive on this. Anything I can help with?

**Reply by davidepianca98** [reply]
Thank you for offering help, but it's actually quite easy. I've released version 0.2.8 containing this improvement, let me know if I missed something.

**Reply by thiagonuic** [reply]
Thanks for the update! I will have a look today and let you know.

**Reply by thiagonuic** [reply]
Hi David,

The current form indeed helps me validate the payload in the authorize part. Since you are already passing username in the `packetReceived` method, would it be possible to also add the password? It would only be available anyway if the constructor flag is set to true anyway. That would allow me to delegate the authorization.

**Reply by davidepianca98** [reply]
Yes I will do that as soon as I can

**Reply by davidepianca98** [reply]
It's now implemented in release 0.2.9

**Reply by thiagonuic** [reply]
Thanks for taking the time to have a look at this use case. Hopefully more people will also benefit from it. I find the username in the authorize specially important in order to give more context.

We are starting a new project on this and as we progress I might come back here to help improve the project.
